oomd: update pressure based kills to use pgscan rate

anitazha opened the pull request:
For pressure based killing we want to target who has the highest
increase in pgscan from the previous interval (vs. the previous logic
which used raw pgscan values). This will prevent biasing towards long
running cgroups as correctly mentioned in #19007.

anitazha commented:
@keszybz would it be possible to do another Fedora systemd release with this PR before the systemd-oomd test week starts this Thursday?

yuwata commented on src/oom/oomd-util.c:
```diff
@@ -413,6 +413,23 @@ int oomd_insert_cgroup_context(Hashmap *old_h, Hashmap *new_h, const char *path)
 }
 
+void oomd_update_cgroup_contexts_between_hashmaps(Hashmap *old_h, Hashmap *curr_h) {
+        _cleanup_(oomd_cgroup_context_freep) OomdCGroupContext *ctx = NULL;
```
`_cleanup_` attribute here seems not necessary.

yuwata commented on src/oom/oomd-util.c:
```diff
@@ -413,6 +413,23 @@ int oomd_insert_cgroup_context(Hashmap *old_h, Hashmap *new_h, const char *path)
+
+        HASHMAP_FOREACH(ctx, curr_h) {
+                OomdCGroupContext *old_ctx;
+                old_ctx = hashmap_get(old_h, ctx->path);
```
Not a matter, but I'd like to add a blank line after the variable declaration.

yuwata commented on src/oom/oomd-util.c:
```diff
@@ -413,6 +413,23 @@ int oomd_insert_cgroup_context(Hashmap *old_h, Hashmap *new_h, const char *path)
+                OomdCGroupContext *old_ctx;
+                old_ctx = hashmap_get(old_h, ctx->path);
+                if (old_ctx) {
+                        ctx->last_pgscan = old_ctx->pgscan;
```
Hm? Question: Is it a typo of `ctx->pgscan`??

yuwata replied:
OK. Thanks.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -232,6 +232,24 @@ static int get_monitored_cgroup_contexts_candidates(Hashmap *monitored_cgroups,
+        _cleanup_hashmap_free_ Hashmap *new_candidates = NULL;
+        int r;
+
+        assert(monitored_cgroups);
```
Maybe, it is better to add two more assertions: `assert(candidates); assert(*candidates);`

yuwata commented on src/oom/oomd-util.h:
```diff
@@ -76,7 +76,12 @@ static inline int compare_pgscan_and_memory_usage(OomdCGroupContext * const *c1,
                 return r;
 
-        r = CMP((*c2)->pgscan, (*c1)->pgscan);
+        /* compare possible underflows first */
+        r = CMP((*c2)->pgscan > (*c2)->last_pgscan, (*c1)->pgscan > (*c1)->last_pgscan);
+        if (r != 0)
+                return r;
+
+        r = CMP((*c2)->pgscan - (*c2)->last_pgscan, (*c1)->pgscan - (*c1)->last_pgscan);
```
Hm, `UINT64_MAX -> 1000` is ordered after `0 -> 1`. Is it expected behavior?

anitazha replied:
Can you elaborate? I'm trying to sort the differences from large to small so I think the first `CMP` will sort the one with subtraction underflow at the end, and this second `CMP` will sort the subtracted difference as expected.

yuwata replied:
When,
```
c2->pgscan = 1000;
c2->last_pgscan = UINT64_MAX;
c1->pgscan = 1;
c1->last_pgscan = 0;
```
then, `c1` is ordered earlier, no? If it is true, is it expected?

anitazha replied:
Oh yup this is expected since essentially c1 increased pgscan but c2 decreased.

yuwata replied:
I see, thanks. Then, one more question. What should we do when
```
c2->pgscan = 500;
c2->last_pgscan = 1000;
c1->pgscan = 200;
c1->last_pgscan = 1000;
```
i.e., pgscan of both contexts decrease. Is it not necessary to handle like the following?
```
r = CMP((*c2)->pgscan > (*c2)->last_pgscan, (*c1)->pgscan > (*c1)->last_pgscan);
if (r != 0)
        return r;

if ((*c2)->pgscan > (*c2)->last_pgscan)
        r = CMP((*c2)->pgscan - (*c2)->last_pgscan, (*c1)->pgscan - (*c1)->last_pgscan);
else
        r = CMP((*c1)->last_pgscan - (*c1)->pgscan, (*c2)->last_pgscan - (*c1)->pgscan);
if (r != 0)
        return r;
```

benzea replied:
I would think something like the following:
```
uint64_t pgscan1, pgscan2;

if ((*c1)->pgscan <= (*c1)->last_pgscan)
  pgscan1 = (*c1)->pgscan - (*c1)->last_pgscan;
else
  pgscan1 = UINT64_MAX - (*c1)->last_pgscan + (*c1)->pgscan;

if ((*c2)->pgscan <= (*c2)->last_pgscan)
  pgscan2 = (*c2)->pgscan - (*c2)->last_pgscan;
else
  pgscan2 = UINT64_MAX - (*c2)->last_pgscan + (*c2)->pgscan;

r = CMP(pgscan1, pgscan2);
if (r != 0)
        return r;
```

The fact that an overflow happened does not really tell us anything about the number of pgscan tables.

poettering replied:
yeah, the right approach with overflows is to have code paths that handle them properly, not just check they exist, i.e. please make sure that for the different overflow conditions there's a codepath that compares values properly

benzea commented:
I wonder if we could avoid updating the information each interval and instead only start when the memory pressure is already above a certain threshold. That would avoids polling a lot of cgroups, but will likely result in having to wait a bit before taking action (full update interval, fixed time?) if the information was not collected in the last interval update.

poettering commented on src/oom/oomd-util.c:
```diff
@@ -413,6 +413,24 @@ int oomd_insert_cgroup_context(Hashmap *old_h, Hashmap *new_h, const char *path)
+                OomdCGroupContext *old_ctx;
+
+                old_ctx = hashmap_get(old_h, ctx->path);
+                if (old_ctx) {
```
doesn't matter, but in the interest of keeping indentation levels short I'd reverse this check to be:

```c
if (!old_ctx)
        continue;
```

i.e. "early exit" is a bit nicer than "deep nesting" I thinkg

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -232,6 +232,25 @@ static int get_monitored_cgroup_contexts_candidates(Hashmap *monitored_cgroups,
+
+        r = get_monitored_cgroup_contexts_candidates(monitored_cgroups, &new_candidates);
+        if (r == -ENOMEM)
+                return r;
```
what's with the other error conditions? are they knowinly ignored? if so, add a comment. or better: debug log about this, saying ", ignoring: %m" in the message

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -297,6 +316,11 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
+        r = update_monitored_cgroup_contexts_candidates(
+                        m->monitored_mem_pressure_cgroup_contexts, &m->monitored_mem_pressure_cgroup_contexts_candidates);
+        if (r == -ENOMEM)
+                return log_error_errno(r, "Failed to update monitored memory pressure candidate cgroup contexts");
```
should have `: %m`

also, wat about the other error again?

poettering reviewed (changes requested):
there's something fishy in the error handling, and that already exists in the code before this PR. This should be cleaned up, i.e. we should make sure that every error is logged once, or ignored explicitly, but never implicitly, because that looks like a bug usuallly.

anitazha commented:
I added another commit to fix the error handling with debug logs and "%m" where necessary.
I also addressed the overflow using something similar to Yu's method. Unit tests with updated with 2 decreasing, 2 increasing, and 1 zeroth pgscan rate change.

anitazha commented:
@benzea I definitely want to optimize this path more since historically polling every second can lead to CPU util of 0.6% - 2% of a core. But probably not in this PR since I want to focus on getting it merged first.

yuwata commented on src/oom/oomd-util.h:
```diff
@@ -76,7 +76,17 @@ static inline int compare_pgscan_and_memory_usage(OomdCGroupContext * const *c1,
+                return r;
+
+        /* If both increased, sort the larger increase first */
+        if ((*c2)->pgscan > (*c2)->last_pgscan && (*c1)->pgscan > (*c1)->last_pgscan)
```
It is not necessary to check pgscan of both contexts are increased.
```
if ((*c2)->pgscan > (*c2)->last_pgscan)
```

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -116,7 +116,8 @@ static int process_managed_oom_reply(
                         r = ret;
                         goto finish;
-                }
+                } else if (ret < 0 && ret != -EEXIST)
```
`else` is redundant.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -156,7 +157,11 @@ static int recursively_get_cgroup_context(Hashmap *new_h, const char *path) {
-                return (r == -ENOMEM) ? r : 0;
+                if (r == -ENOMEM)
+                        return r;
+                else if (r < 0)
```
Here too. `else` is redundant.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -171,15 +176,21 @@ static int recursively_get_cgroup_context(Hashmap *new_h, const char *path) {
-                        return (r == -ENOMEM) ? r : 0;
+                if (r == -ENOMEM)
+                        return r;
+                else if (r < 0) {
```
Also here. `else` is redundant.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -171,15 +176,21 @@ static int recursively_get_cgroup_context(Hashmap *new_h, const char *path) {
                         r = recursively_get_cgroup_context(new_h, cg_path);
                 if (r == -ENOMEM)
                         return r;
+                else if (r < 0)
```
Here too.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -227,6 +238,8 @@ static int get_monitored_cgroup_contexts_candidates(Hashmap *monitored_cgroups,
                 r = recursively_get_cgroup_context(candidates, ctx->path);
                 if (r == -ENOMEM)
                         return r;
+                else if (r < 0)
```
Here too.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -298,38 +311,44 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
         if (r == -ENOMEM)
-                return log_error_errno(r, "Failed to update monitored swap cgroup contexts");
+                return log_error_errno(r, "Failed to update monitored swap cgroup contexts: %m");
+        else if (r < 0)
```
Here too.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -298,38 +311,44 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
         if (r == -ENOMEM)
-                return log_error_errno(r, "Failed to update monitored memory pressure cgroup contexts");
+                return log_error_errno(r, "Failed to update monitored memory pressure cgroup contexts: %m");
+        else if (r < 0)
```
Here too.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -298,38 +311,44 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
         if (r == -ENOMEM)
-                return log_error_errno(r, "Failed to update monitored memory pressure candidate cgroup contexts");
+                return log_error_errno(r, "Failed to update monitored memory pressure candidate cgroup contexts: %m");
+        else if (r < 0)
```
Here too.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -346,7 +365,9 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
         if (r == -ENOMEM)
-                return log_error_errno(r, "Failed to check if memory pressure exceeded limits");
+                return log_error_errno(r, "Failed to check if memory pressure exceeded limits: %m");
+        else if (r < 0)
```
Here too.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -382,13 +403,15 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
                 if (r == -ENOMEM)
-                        return log_error_errno(r, "Failed to get monitored swap cgroup candidates");
+                        return log_error_errno(r, "Failed to get monitored swap cgroup candidates: %m");
+                else if (r < 0)
```
Here too.

yuwata reviewed (changes requested):
Mostly good. But there are many redundant `else`.

anitazha commented:
Thanks yuwata I pushed an update with your suggestions.

yuwata commented on src/oom/oomd-manager.c:
```diff
@@ -297,6 +320,11 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
+        r = update_monitored_cgroup_contexts_candidates(
+                        m->monitored_mem_pressure_cgroup_contexts, &m->monitored_mem_pressure_cgroup_contexts_candidates);
+        if (r == -ENOMEM)
+                return log_error_errno(r, "Failed to update monitored memory pressure candidate cgroup contexts");
```
Not addressed...

anitazha replied:
I ended up changing `update_monitored_cgroup_contexts_candidates()` to return whenever `r < 0` so I could keep this statement as is.

yuwata replied:
Oops, It seems I looked a cached old commit. Yes. looks good.

yuwata commented:
Looks good, except for one point which is not addressed yet.

bluca reviewed (approved):
LGTM too. The s390x test failure in nspawn seems unrelated.

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -379,13 +403,15 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
+                        return log_error_errno(r, "Failed to kill cgroup processes by swap usage: %m");
                 if (r < 0)
-                        log_info("Failed to kill any cgroup(s) based on swap");
+                        log_info_errno(r, "Failed to kill any cgroup(s) based on swap: %m");
```
probably should be "log_notice_errno()", i.e. this is more than informational, no? also should say ", ignoring: %m" I figure

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -361,7 +385,7 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
                                         return log_error_errno(r, "Failed to kill cgroup processes by pgscan rate: %m");
                                 if (r < 0)
-                                        log_info("Failed to kill any cgroup(s) under %s based on pressure", t->path);
+                                        log_info_errno(r, "Failed to kill any cgroup(s) under %s based on pressure: %m", t->path);
```
as below

poettering reviewed (commented):
just nitpicks from my side which don#t really matter...

anitazha commented on src/oom/oomd-util.h:
```diff
@@ -76,7 +76,17 @@ static inline int compare_pgscan_and_memory_usage(OomdCGroupContext * const *c1,
 
-        r = CMP((*c2)->pgscan, (*c1)->pgscan);
+        /* If one increased and the other decreased, sort the increased rate first */
+        r = CMP((*c2)->pgscan > (*c2)->last_pgscan, (*c1)->pgscan > (*c1)->last_pgscan);
```
I see what you mean now. I confirmed with the kernel folks that lest there be a kernel bug or the cgroup was recreated pgscan should be monotonic. So I think what I want to do here is it (pgscan > last_pgscan), assume like last_pgscan was zero. and that the cgroup was recreated. But also log about it.

benzea replied:
Ohh, good point on the cgroup being possibly recreated!

bluca replied:
> In practice the kernel shouldn't overflow unless "70 zettabytes worth of memory" were scanned.

So it will be fine as long as one doesn't open more than 3 tabs in Chrome, in other words :-D

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -275,33 +311,44 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
         r = update_monitored_cgroup_contexts(&m->monitored_swap_cgroup_contexts);
         if (r == -ENOMEM)
-                return log_error_errno(r, "Failed to update monitored swap cgroup contexts");
+                return log_error_errno(r, "Failed to update monitored swap cgroup contexts: %m");
```
given this is just the OOM condition please use:

```c
return log_oom();
```

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -275,33 +311,44 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
         r = update_monitored_cgroup_contexts(&m->monitored_mem_pressure_cgroup_contexts);
         if (r == -ENOMEM)
-                return log_error_errno(r, "Failed to update monitored memory pressure cgroup contexts");
+                return log_error_errno(r, "Failed to update monitored memory pressure cgroup contexts: %m");
```
ditto

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -318,30 +365,27 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
         r = oomd_pressure_above(m->monitored_mem_pressure_cgroup_contexts, m->default_mem_pressure_duration_usec, &targets);
         if (r == -ENOMEM)
-                return log_error_errno(r, "Failed to check if memory pressure exceeded limits");
+                return log_error_errno(r, "Failed to check if memory pressure exceeded limits: %m");
```
ditto

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -318,30 +365,27 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
+                                r = oomd_kill_by_pgscan_rate(m->monitored_mem_pressure_cgroup_contexts_candidates, t->path, m->dry_run);
                                 if (r == -ENOMEM)
-                                        return log_error_errno(r, "Failed to kill cgroup processes by pgscan");
+                                        return log_error_errno(r, "Failed to kill cgroup processes by pgscan rate: %m");
```
ditto

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -359,13 +403,15 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
                 r = get_monitored_cgroup_contexts_candidates(m->monitored_swap_cgroup_contexts, &candidates);
                 if (r == -ENOMEM)
-                        return log_error_errno(r, "Failed to get monitored swap cgroup candidates");
+                        return log_error_errno(r, "Failed to get monitored swap cgroup candidates: %m");
```
ditto

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -359,13 +403,15 @@ static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, vo
                 r = oomd_kill_by_swap_usage(candidates, m->dry_run);
                 if (r == -ENOMEM)
-                        return log_error_errno(r, "Failed to kill cgroup processes by swap usage");
+                        return log_error_errno(r, "Failed to kill cgroup processes by swap usage: %m");
```
ditto

poettering reviewed (changes requested):
one more simple fix please

(yeah, i know I am nitpicking)

poettering commented:
thanks! let's get this merged!
